Tests/AlicerceTests/Extensions/Foundation/StringTestCase.swift

```diff
@@ -66,5 +66,110 @@ class StringTestCase: XCTestCase {
+
+        let original = "0123456789ABCDEF"
+        let expected = "0123456789abcdef"
+
+        XCTAssertEqual(
+            original.replacingOccurrencesOfCharacters(
+                in: [
+                    .init("A"): "a",
+                    .init("B"): "b",
+                    .init("C"): "c",
+                    .init("D"): "d",
+                    .init("E"): "e",
+                    .init("F"): "f",
+                ],
+                skippingCharactersIn: nil
+            ),
+            expected
+        )
+    }
+
+    func testReplacingOccurrencesOfCharacters_WithMatchingCharactersInMapAndMatchingCharactersInSkippingCharacterSet_ShouldReplaceOccurrencesAndSkip() {
+
+        let original = "0123456789ABCDEF_0A0B0C0D0E0F0"
+        let expected = "abcdef_abcdef"
+
+        XCTAssertEqual(
+            original.replacingOccurrencesOfCharacters(
+                in: [
+                    .init("A"): "a",
+                    .init("B"): "b",
+                    .init("C"): "c",
+                    .init("D"): "d",
+                    .init("E"): "e",
+                    .init("F"): "f",
+                ],
+                skippingCharactersIn: .decimalDigits
+            ),
+            expected
+        )
+    }
+
+    // nonLineBreaking()
+
+    func testNonLineBreaking_WithNoLineBreakingCharactersInString_ShouldReturnSelf() {
+
+        let original = "0123456789ABCDEF"
+
+        XCTAssertEqual(original.nonLineBreaking(), original)
+    }
+
+    func testNonLineBreaking_WithLineBreakingCharactersInString_ShouldReturnANonLineBreakingVersion() {
+
+        let original = "The quick-brown\(String.emDash)fox\(String.enDash)jumps?over{the}lazy dog"
+        let expected =
+            """
+            The\(String.nonBreakingSpace)quick\(String.nonBreakingHyphen)brown\
+            \(String([.wordJoiner, .emDash, .wordJoiner]))fox\
+            \(String([.wordJoiner, .enDash, .wordJoiner]))jumps\
+            ?\(String.wordJoiner)over{the}\(String.wordJoiner)lazy\(String.nonBreakingSpace)dog
+            """
+
+        XCTAssertEqual(original.nonLineBreaking(), expected)
+    }
+
+    func testNonLineBreaking_WithLineBreakingCharactersAndNewlinesInString_ShouldReturnANonLineBreakingVersion() {
+
+        let original =
+            """
+            \nThe quick-brown\u{85}\(String.emDash)fox\n\(String.enDash)jumps?\u{2028}\u{2029}over{the}lazy dog\n\
+            \u{A}.\u{B},\u{C};\u{D}
+            """
+        
+        let expected =
+            """
+            \nThe\(String.nonBreakingSpace)quick\(String.nonBreakingHyphen)brown\u{85}\
+            \(String([.wordJoiner, .emDash, .wordJoiner]))fox\n\
+            \(String([.wordJoiner, .enDash, .wordJoiner]))jumps\
+            ?\(String.wordJoiner)\u{2028}\u{2029}over\
+            {the}\(String.wordJoiner)lazy\(String.nonBreakingSpace)dog\n\
+            \u{A}.\u{B},\u{C};\u{D}
+            """
+
+        XCTAssertEqual(original.nonLineBreaking(), expected)
+    }
```

Comment:
The semantics of this test are a bit misleading 😅 as `\n` is also known as a "line break" but they are expected to be preserved when calling `.nonLineBreaking()` 😵 . I've tried and failed to come up with an alternative method name 😞 especially since upon futher inspection the function documentation clearly states which characters are considered "line breaking" 🤷

Reply:
This test was initially designed to test whether newlines were skipped or not, but after some iterations it reached this form where it tests that they're actually preserved. 🙈 I'll rephrase the test name to make it a bit clearer.

The main purpose of the `nonLineBreaking()` helper is to avoid breaking text where it's usually broken when rendered by the system and you don't "control it" - thus we "force" it to via non breaking characters.  It's not to make the string a single "unbreaked" line (which would then require us to convert/strip newlines too). 

As mentioned in a [previous discussion](https://github.com/Mindera/Alicerce/pull/227#discussion_r607202291), automatically converting newlines is not straightforward, so it's safer to preserve them by default. As an escape hatch, we provide the "raw" `replacingOccurrencesOfCharacters` so that users can do whatever they want. The method has been further documented to clearly state that newlines are preserved.

Hope this helped clarify the reasoning. 😄 

Do you have any concrete suggestions?

Reply:
one of the names that came to mind was `reduceUnexpectedLineBreaks()` but I'm the first to admit that it's not a very good name, so I'm perfectly content with your explanation 🙈

Reply:
well, we could add in an extra knob (parameter), which would allow callers to choose what substitution to use for newlines. Something like:

`public func nonLineBreaking(newlineCharacterReplacement: String? = nil) -> String`

Which unlocks the following capabilities depending on the `newlineCharacterReplacement` value:
- `nil`: do nothing - newlines are preserved
- `""`: strip newlines
- `String.nonBreakingSpace`: one line to rule them all
- `"🍌"`: make monkey users happy

I guess that this improves the current API by tying these "loose ends" and giving even more control.

Reply:
Bonus points 💯⭐ for the default value that preserves the current calls 👍